curiosrc/alertmanager/alerts.go

```diff
@@ -0,0 +1,288 @@
+	tmap := make(map[string]int)
+
+	if len(taskFailures) > 0 {
+		for _, tf := range taskFailures {
+			_, ok := tmap[tf.Name]
+			if !ok {
+				tmap[tf.Name] = tf.Failures
+			} else {
+				tmap[tf.Name] += tf.Failures
+			}
+			_, ok = mmap[tf.Machine]
+			if !ok {
+				mmap[tf.Machine] = tf.Failures
+			} else {
+				mmap[tf.Machine] += tf.Failures
+			}
+		}
+	}
+
+	sealingTasks := []string{"SDR", "TreeD", "TreeRC", "PreCommitSubmit", "PoRep", "Finalize", "MoveStorage", "CommitSubmit", "WdPost", "ParkPiece"}
+	contains := func(s []string, e string) bool {
+		for _, a := range s {
+			if a == e {
+				return true
+			}
+		}
+		return false
+	}
+
+	// Alerts for any sealing pipeline failures. Other tasks should have at least 5 failures for an alert
+	for name, count := range tmap {
+		if contains(sealingTasks, name) {
+			al.alertMap[Name].alertString += fmt.Sprintf("Task: %s, Failures: %d. ", name, count)
+		}
+		if count > 5 {
+			al.alertMap[Name].alertString += fmt.Sprintf("Task: %s, Failures: %d. ", name, count)
+		}
+	}
+
+	// Alert if a machine failed more than 5 tasks
+	for name, count := range tmap {
+		if count > 5 {
+			al.alertMap[Name].alertString += fmt.Sprintf("Machine: %s, Failures: %d. ", name, count)
+		}
+	}
+
+	return
+}
+
+// permanentStorageCheck checks the available storage space for sealing sectors.
+// It retrieves the storage paths and the sectors being sealed from the database.
+// It calculates the total required storage space by summing the sector sizes of the sectors being sealed.
+// It compares the total required space with the total available space.
+func permanentStorageCheck(al *alerts) {
+	Name := "PermanentStorageSpace"
+	// Get all storage path for permanent storages
+	type storage struct {
+		ID        string `db:"storage_id"`
+		Available int64  `db:"available"`
+	}
+
+	var storages []storage
+
+	err := al.db.Select(al.ctx, &storages, `
+								SELECT storage_id, available
+								FROM storage_path
+								WHERE can_store = TRUE;`)
+	if err != nil {
+		al.alertMap[Name].err = xerrors.Errorf("getting storage details: %w", err)
+		return
+	}
+
+	type sector struct {
+		Miner  abi.ActorID             `db:"sp_id"`
+		Number abi.SectorNumber        `db:"sector_number"`
+		Proof  abi.RegisteredSealProof `db:"reg_seal_proof"`
+	}
+
+	var sectors []sector
+
+	err = al.db.Select(al.ctx, &sectors, `
+								SELECT sp_id, sector_number, reg_seal_proof
+								FROM sectors_sdr_pipeline
+								WHERE after_finalize = FALSE;`)
+	if err != nil {
+		al.alertMap[Name].err = xerrors.Errorf("getting sectors being sealed: %w", err)
+		return
+	}
+
+	var totalRequiredSpace int64
+	for _, sec := range sectors {
+		sec := sec
+		sectorSize, err := sec.Proof.SectorSize()
+		if err != nil {
+			totalRequiredSpace += int64(64 << 30)
+		}
+		totalRequiredSpace += int64(sectorSize)
+	}
+
+	var totalAvailableSpace int64
+	for _, storage := range storages {
+		totalAvailableSpace += storage.Available
+	}
+
```

Comment:
totalRequiredSpace += largestSectorSize * len(storages)
 // without this, bin packing differences and half-sector aggregations would make this thing never alert.

Reply:
That could be too much (especially if every hard drive is a storage). 
Your alternative would be to do bin-packing by remembering how many of each sector size, and burn-down the list with some arrangement.

Reply:
This was poor logic. I have updated the code block to account for each sector individually on storages. If all sectors are accounted for then no alerts but if we have a sector which could not be places anywhere, then an alert is created.